portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java

```diff
@@ -0,0 +1,171 @@
+      // no merging or adding needs to happen.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        continue;
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Change the lastExaminedTimeRange pointer to the new time range created
+      // after the merging.
+      if (lastExaminedTimeRange.overlaps(currentRange)) {
+        lastExaminedTimeRange = mergeTwoTimeRanges(currentRange, lastExaminedTimeRange);
+      } else {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list of all ranges, and
+        // change the lastExaminedTimeRange to point to the current range.
+        if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+          newTimeRanges.add(lastExaminedTimeRange);
+          lastExaminedTimeRange = currentRange;
+        } else {
+          // If lastExaminedTimeRange is later than the current range,
+          // add the current range to the new list of all ranges.
+          // The lastExaminedTimeRange pointer shouldn't move.
+          newTimeRanges.add(currentRange);
+        }
+      }
+
+      // If current time range is the last element in the allTimeRanges list,
+      // add the last examined time range to the new list.
+      if (i == allTimeRanges.size() - 1) {
+        newTimeRanges.add(lastExaminedTimeRange);
+      }
+    }
+
+    // Finally, set the global variable allTimeRanges to this newly built list of time ranges.
+    allTimeRanges = newTimeRanges;
+    Collections.sort(allTimeRanges, TimeRange.SORT_BY_TIME_RANGE_START_TIME_ASCENDING);
+  }
+
+  /** Helper method for merging two time ranges. This method is public so that it can be tested. */
+  public static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
+    Instant newTimeRangeStart;
+    Instant newTimeRangeEnd;
+
+    // The new time range after merging should have the earlier start time
+    // and the later end time among the two overlapping time ranges.
+    if (a.start().isBefore(b.start())) {
+      newTimeRangeStart = a.start();
+    } else {
+      newTimeRangeStart = b.start();
+    }
+
+    if (a.end().isBefore(b.end())) {
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> getAllTimeRanges() {
+    return allTimeRanges.iterator();
+  }
+
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
```

Comment:
It'd be nice to use binary search for a method like this in the future, if you can.

Reply:
Updated.